src/pineko/kfactor.py

```diff
@@ -209,172 +168,133 @@ def create_singlegridonly(
-        # this is not a problem because in those cases either all cfactor values are the
-        # same (thus there is no doubt about whether we have the correct one) or the
-        # non-exisiting cfactors would be multiplied by bins corresponding to all '0' in the
-        # grid.
-        # Let's check if we are in the first or second case
-        if len(np.unique(centrals_kfactor)) == 1:
-            # In this case I just need to add more elements to the kfactor
-            for _num in range(pigrid.bins()):
-                centrals_kfactor_filtered = np.append(
-                    centrals_kfactor_filtered, centrals_kfactor[0]
-                )
-        else:
-            # In this case this means that the missing entries will multiply zero subgrids so we can just add 0s
-            for _num in range(pigrid.bins()):
-                centrals_kfactor_filtered = np.append(centrals_kfactor_filtered, 0.0)
-    return centrals_kfactor_filtered
 
+def to_list(grid, central_kfactors):
+    """Cast the centrals kfactors to the correct length.
 
-def compute_k_factor_grid(
+    Apply a normalization according to the length compared to the number of bins of the grid.
+
+    Parameters
+    ----------
+    grid: pineappl.grid.Grid
+        grid
+    centrals_kfactor: list
+        list of kfactor for each point
+    """
+    if grid.bins() == len(central_kfactors):
+        rich.print("The number of bins match the length of the kfactor.")
+        return central_kfactors
+    if grid.bins() < len(central_kfactors):
+        rich.print(
+            "[yellow] The number of bins is less than the length of the kfactor."
+        )
+        if not len(np.unique(central_kfactors)) == 1:
+            # This case is actually wrong.
+            raise ValueError("kfactor contains too many different values.")
+        return central_kfactors
+
+    rich.print("[yellow] The number of bins is more than the length of the kfactor.")
+
+    # This is the last case in which grid.bins() > len(centrals_kfactor)
+    # Note that sometimes there are more bins in the grid than in the kfactor file -
+    # this is not a problem because in those cases either all kfactor values are the
+    # same (thus there is no doubt about whether we have the correct one) or the
+    # non-exisiting kfactor would be multiplied by bins corresponding to all '0' in the
+    # grid.
+    # Let's check if we are in the first or second case
+    if len(np.unique(central_kfactors)) == 1:
+        # In this case I just need to add more elements to the kfactor
+        return np.full(grid.bins(), central_kfactors[0])
+    # In this case this means that the missing entries will
+    # multiply zero subgrids so we can just add 0s
+    return np.full(grid.bins(), 0)
+
+
+def apply_to_dataset(
     grids_folder,
     kfactor_folder,
     yamldb_path,
-    max_as,
+    pto_to_update,
     target_folder=None,
```

Comment:
```suggestion
    target_folder,
```
I don't think `None` works - specifically `target_folder / grid_name` does not

Reply:
Yes, let's drop it. I believe it' better to have it always specified and not overwrite the previous grid.

Reply:
done in [25be7c4](https://github.com/NNPDF/pineko/pull/161/commits/25be7c4c1df7fab88905c5c2fc788acb39dd3d80)